**Situation:**

This script matches IATA codes / OAG ac names and Planespotters ac.

There is a general difference between the IATA and the ICAO aircraft codes: IATA has designated, superordinated code for "all models" while the ICAO has not. (In the following, we call these aircraft types "IATA-only".)

This becomes clear at the following example:
- IATA=141 for B461 BAe 146-100 Pax --> ICAO=B464
- IATA=142 for B462 BAe 146-200 Pax --> ICAO=B462
- IATA=143 for B463	BAe 146-300 Pax --> ICAO=B463
- IATA=146 for BAe 146 all pax models --> ICAO=N/A <- this is the general, superordinated IATA code.

**Problem**: 

This OAG data set is based on the IATA code, but the Planespotters (PS) data (data on aircraft, including e.g. age and seat capacity) does always have specific AC models, not the general, superordinated AC.

**Solution**:

We distribute the OAG flights performed by the IATA-only AC types to all specific IATA-code AC that are available in the set. We do so per route and according to the share of flight kilometers observed in the OAG data.

Example for 1 specific route:
- IATA=146: 10 flights & 2000 FKM in OAG data set --> 0 FKM after reassignment
- IATA=141: 20 flights & 4000 FKM in OAG data set --> 4800 FKM after reassignment
- IATA=142: 30 flights & 6000 FKM in OAG data set --> 7200 FKM after reassignment
The 2000 FKM from IATA=146 are reassigned to the 141 and 142 AC according to the share of FKM in OAG.

In [1]:
import pandas as pd
import numpy as np
import math

# 2018

## (2018) Load OAG data

In [2]:
# Connect to DB
%load_ext ipython_pg.LAV
%pg_connect dbname='mobility' host='lav-fileserver.ethz.ch' user='heldm'

password for heldm@lav-fileserver.ethz.ch:5433:········
SUCCESS: connected to lav-fileserver.ethz.ch
  PostGIS integration enabled

In [3]:
%%pg_pd oag18
SELECT * FROM "OAG"."2018_schedule"

SUCCES: matched 180578 rows
 results stored as 'oag18'


In [4]:
oag18.head()

ac_code_iata ac_name_oag arr_ap_code   arr_ap_name arr_ap_cty dep_ap_code  \
0          100  Fokker 100         ADE          Aden         YE         JIB   
1          100  Fokker 100         AMS     Amsterdam         NL         BRU   
2          100  Fokker 100         AMS     Amsterdam         NL         MUC   
3          100  Fokker 100         AZD          Yazd         IR         BND   
4          100  Fokker 100         BND  Bandar Abbas         IR         AZD   

                    dep_ap_name dep_ap_cty  gc_distance  frequency_2018  \
0                      Djibouti         DJ   248.203212            58.0   
1              Brussels Airport         BE   157.553624            88.0   
2  Munich International Airport         DE   676.777570           332.0   
3                  Bandar Abbas         IR   559.459639            91.0   
4                          Yazd         IR   559.459639            91.0   

        fkm_2018  
0   14395.786303  
1   13864.718918  
2  224690.153347  
3   50910.827163  
4   50910.827163

## (2018) Make OAG data set Planespotters-compatible

### (2018) Reassign generic IATA Codes

**Load manual IATA-ICAO lookup table**

In [5]:
# Read the manual lookup table
ac_type_designators = pd.read_excel('../../../../Research/Aviation/Jupyter/datasets/aircraft_type_designators_manual.xlsx')
ac_type_designators = ac_type_designators.rename(columns={'oag_model: all generic AC types where we have multiple variants in the PS data set are marked in red': 'ac_name_oag',
                                            'planespotters_model': 'ac_name_ps',
                                            'share of yearly flight km: high impact models (>1% of total FKM) are marked in green -> ask OAG here': 'share_of_yearly_FKM'})
ac_type_designators = ac_type_designators.drop(['share_of_yearly_FKM','ps_model assignment','comment1','comment2', 'widebody'], axis=1)
ac_type_designators['ac_code_iata_reassign'] = ac_type_designators['ac_code_iata_reassign'].str.split(',')
ac_type_designators['ac_code_iata'] = ac_type_designators['ac_code_iata'].astype(str)
ac_type_designators.head()

ac_code_iata ac_code_icao                 ac_name_oag       ac_name_ps  \
0          CS1         BCS1             Airbus A220-100  Airbus A220-100   
1          CS3         BCS3             Airbus A220-300  Airbus A220-300   
2          AB3         A30B       Airbus A300 Passenger    Airbus A300B4   
3          AB6         A306   Airbus A300-600 Passenger  Airbus A300-600   
4          AB4         A30B  Airbus A300B2/B4 Passenger    Airbus A300B4   

  ac_code_iata_reassign               oag_name_keep  
0                   NaN                         NaN  
1                   NaN                         NaN  
2                   NaN                          no  
3                   NaN                         NaN  
4                   NaN  Airbus A300B2/B4 Passenger

In [6]:
# Extract all iata codes that have to be reassigned
reassign_df = ac_type_designators[ac_type_designators['ac_name_ps'] == 'to_be_assigned_to_multiple_variants'].reset_index()
reassign_df

index ac_code_iata                         ac_code_icao  \
0       8          32S  to_be_assigned_to_multiple_variants   
1      16          330  to_be_assigned_to_multiple_variants   
2      20          340  to_be_assigned_to_multiple_variants   
3      25          350  to_be_assigned_to_multiple_variants   
4      40          ATR  to_be_assigned_to_multiple_variants   
5      42          ARJ  to_be_assigned_to_multiple_variants   
6      44          146  to_be_assigned_to_multiple_variants   
7      62          M80  to_be_assigned_to_multiple_variants   
8      73          737  to_be_assigned_to_multiple_variants   
9      98          757  to_be_assigned_to_multiple_variants   
10    103          767  to_be_assigned_to_multiple_variants   
11    108          777  to_be_assigned_to_multiple_variants   
12    113          787  to_be_assigned_to_multiple_variants   
13    120          CRJ  to_be_assigned_to_multiple_variants   
14    141          DH8  to_be_assigned_to_multiple_variants   
15    149          EMJ  to_be_assigned_to_multiple_variants   
16    156          ERJ  to_be_assigned_to_multiple_variants   

                      ac_name_oag                           ac_name_ps  \
0         Airbus A318/319/320/321  to_be_assigned_to_multiple_variants   
1                     Airbus A330  to_be_assigned_to_multiple_variants   
2                     Airbus A340  to_be_assigned_to_multiple_variants   
3                     Airbus A350  to_be_assigned_to_multiple_variants   
4                    ATR42 /ATR72  to_be_assigned_to_multiple_variants   
5          Avro RJ70 /rj85 /rj100  to_be_assigned_to_multiple_variants   
6               BAe 146 Passenger  to_be_assigned_to_multiple_variants   
7          Boeing (douglas) MD-80  to_be_assigned_to_multiple_variants   
8            Boeing 737 Passenger  to_be_assigned_to_multiple_variants   
9          Boeing 757 (Passenger)  to_be_assigned_to_multiple_variants   
10           Boeing 767 Passenger  to_be_assigned_to_multiple_variants   
11           Boeing 777 Passenger  to_be_assigned_to_multiple_variants   
12                     Boeing 787  to_be_assigned_to_multiple_variants   
13          Canadair Regional Jet  to_be_assigned_to_multiple_variants   
14  DHvilld-Bombardier DHC8 Dsh 8  to_be_assigned_to_multiple_variants   
15                Embraer 170/195  to_be_assigned_to_multiple_variants   
16         Embraer RJ 135/140/145  to_be_assigned_to_multiple_variants   

                                ac_code_iata_reassign oag_name_keep  
0                      [318, 319, 320, 32A, 321, 32B]           NaN  
1                                          [332, 333]           NaN  
2                                [342, 343, 345, 346]           NaN  
3                                          [351, 359]           NaN  
4                                          [AT4, AT7]           NaN  
5                                          [AR1, AR8]           NaN  
6                                     [141, 142, 143]           NaN  
7                           [M81, M82, M83, M87, M88]           NaN  
8   [732, 73C, 733, 734, 73E, 735, 736, 73W, 73G, ...           NaN  
9                                          [752, 753]           NaN  
10                                    [762, 763, 764]           NaN  
11                                         [772, 773]           NaN  
12                                    [781, 788, 789]           NaN  
13                     [CR1, CR2, CR7, CR9, CRA, CRK]           NaN  
14                               [DH1, DH2, DH3, DH4]           NaN  
15                                         [E70, E95]           NaN  
16                                    [ER3, ER4, ERD]           NaN

**Reassign**

In [7]:
# Add the total share of yearly fkm of each OAG entry and a one-string-route as two new columns
oag18['share_of_fkm_2018'] = oag18['fkm_2018'] / oag18['fkm_2018'].sum()
oag18['route'] = oag18['arr_ap_code'] + '_' + oag18['dep_ap_code']
oag18.head()

ac_code_iata ac_name_oag arr_ap_code   arr_ap_name arr_ap_cty dep_ap_code  \
0          100  Fokker 100         ADE          Aden         YE         JIB   
1          100  Fokker 100         AMS     Amsterdam         NL         BRU   
2          100  Fokker 100         AMS     Amsterdam         NL         MUC   
3          100  Fokker 100         AZD          Yazd         IR         BND   
4          100  Fokker 100         BND  Bandar Abbas         IR         AZD   

                    dep_ap_name dep_ap_cty  gc_distance  frequency_2018  \
0                      Djibouti         DJ   248.203212            58.0   
1              Brussels Airport         BE   157.553624            88.0   
2  Munich International Airport         DE   676.777570           332.0   
3                  Bandar Abbas         IR   559.459639            91.0   
4                          Yazd         IR   559.459639            91.0   

        fkm_2018  share_of_fkm_2018    route  
0   14395.786303       2.767831e-07  ADE_JIB  
1   13864.718918       2.665724e-07  AMS_BRU  
2  224690.153347       4.320044e-06  AMS_MUC  
3   50910.827163       9.788459e-07  AZD_BND  
4   50910.827163       9.788459e-07  BND_AZD

In [8]:
def lookup_fkm_shares(iata_orig, ls_reassign, oag18_route, oag18):
    '''Take the original generic iata code aircraft (iata_orig) that has to be reassigned to other aircraft types,
    which are the input list ls_reassign aircraft types.
    Takes the oag data for a specific route - all aircraft that perform these route.
    Returns relative shares of these to-be-reassigned aircraft types.'''
    
    ls_reassign = pd.DataFrame({'ac_code_iata': ls_reassign, 'share_of_fkm_2018': 'share_of_fkm_2018'})
    ac_performance = oag18.groupby('ac_code_iata').sum().reset_index()
    ls_reassign['share_of_fkm_2018'] = ac_performance.set_index('ac_code_iata').lookup(ls_reassign['ac_code_iata'],ls_reassign['share_of_fkm_2018'])
    ls_reassign['share_of_fkm_2018'] = ls_reassign['share_of_fkm_2018'] / ls_reassign['share_of_fkm_2018'].sum()
    
    return ls_reassign

In [9]:
new_oag18 = oag18.copy()

In [10]:
'''Some iata codes in the OAG data set do not refer to a specific aircraft, but are a general "melting pot" for all aircraft versions of a general type.
Since stock data from Planespotters does refer only to specific aircraft, we have to reassign the general aircraft types to the specific ones.
This code block assures that this reassignment (redistribution of number of flights (frequency_2018) and flight kilometers (fkm_2018)) is done according to 
the share of specific aircraft versions on the total fkm in 2018. Since fkm correlate good with number of aircraft in a first order approximation,
yearly fkm share is a good proxy for the number of aircraft version that we reassign.'''

for iata_orig in np.arange(len(reassign_df)): # loop over all iata codes that have to be replaced (iata_orig = [0,1,2,...,#all orig iata codes])
    
    orig = reassign_df.loc[iata_orig,'ac_code_iata']           # original ac that has to be reassigned (e.g. '330')
    reass = reassign_df.loc[iata_orig,'ac_code_iata_reassign'] # ac to which the fkm are reassigned (e.g. ['332', '333'])
    
    
    for route in oag18[oag18['ac_code_iata'] == orig]['route']: # loop over routes
        
        if len(oag18[(oag18['route'] == route) & (oag18['ac_code_iata'] == orig)]) > 0: # check if orig ac is operated at the current route
            # Look sequentially at routes in the oag data for the orig aircraft
            oag18_route = oag18[(oag18['route'] == route) & \
                                ((oag18['ac_code_iata'] == orig) | (oag18['ac_code_iata'].isin(reass)))].reset_index()
            oag18_route = oag18_route.drop(columns='index')

            if len(oag18_route) > 0:
                # Create a new table for all reass aircraft and set empty cells that are filled afterwards
                orig_oag18_route = oag18_route[oag18_route['ac_code_iata'] == orig]
                new_df = pd.concat([orig_oag18_route]*len(reass), ignore_index=True)
                new_df.at[:,['ac_code_iata', 'ac_name_oag','frequency_2018', 'fkm_2018', 'share_of_fkm_2018']] = ''
                new_df.at[:,'ac_code_iata'] = reass
                new_df.at[:,'ac_name_oag'] = 'new_reass_ac'

                # Define shares of yearly fkm shares of the reass aircraft
                shares = lookup_fkm_shares(orig, reass, oag18_route, oag18)

                # Calculate the fkm that have to be redistributed from the orig ac to the reass ac
                orig_fkm = oag18_route[oag18_route['ac_code_iata'] == orig]['fkm_2018'].item()


                # Loop over all reass ac for the specific aircraft orig (loop 1) and the specific route (loop 2)
                for i in reass:
                    # Get the share of fkm of i-th reass ac on the total fkm of all reass ac
                    share_i = shares[shares['ac_code_iata'] == i]['share_of_fkm_2018'].item()
            
                    # Check whether the i-th reass aircraft is already assigned for that route in the original oag data set and save the fkm of it if it is nonzero.
                    reass_fkm_old = oag18_route[oag18_route['ac_code_iata'] == i]['fkm_2018']
                    if len(reass_fkm_old) >0:
                        reass_fkm_old = reass_fkm_old.item()
                    else:
                        reass_fkm_old = 0
                    
                    # Fill the dataframe for the reass aircraft with the share-wise distribution of the to-be-reassigned orig fkm + the reass_fkm_old
                    row = new_df[new_df['ac_code_iata'] == i].index.item()
                    
                    
                    new_df.at[row,'fkm_2018'] = share_i*orig_fkm + reass_fkm_old
                    # Calculate the frequency from the fkm and the great-circle distance
                    new_df.at[row,'frequency_2018'] = round(new_df.loc[row,'fkm_2018'] / new_df.loc[row,'gc_distance'])   


                # Since we assign the fkm according to yearyl fkm shares, the frequency for a specific route-ac pair might not be a multiple of the gc_distance (see last step in reass-loop).
                # Solution: Make sure that the sum of frequency_2018 over all reass ac is the same as for all ac (orig + reass) on the route.
                diff_freq = new_df['frequency_2018'].sum() - oag18_route['frequency_2018'].sum() 

                # If there is a difference, add or substract that difference in number of flights from the reass ac with the largest freq,
                # unless there is no reass ac with a largest freq, then take the last ac in the list
                idx_max = pd.to_numeric(new_df['frequency_2018'])

                if abs(diff_freq) > 0:
                    if math.isnan(idx_max.idxmax()):
                        last_row = new_df.index[-1]
                        new_df.at[last_row,'frequency_2018'] -= diff_freq
                    else:
                        new_df.at[idx_max.idxmax(),'frequency_2018'] -= diff_freq

                # After this correction, we can reset the fkm's according to the frequencies:
                new_df.at[:,'fkm_2018'] = new_df.loc[:,'frequency_2018'] * new_df.loc[:,'gc_distance']

                # Now we append the rows with the new values for the reass ac and drop the original ones if existing
                new_oag18 = new_oag18.append(new_df, ignore_index=True)
                new_oag18 = new_oag18.drop_duplicates(subset = ['ac_code_iata','route'], keep = 'last')
                
    # Last we delete all orig ac
    idx_to_del = new_oag18[new_oag18['ac_code_iata'] == orig].index
    new_oag18 = new_oag18.drop(index=idx_to_del)

In [11]:
# Since the fkm-frequency-step in the loop above creates a few 0-frequency-entries, we delete them.
zeros = new_oag18[new_oag18['frequency_2018'] == 0].index
new_oag18 = new_oag18.drop(index=zeros)
new_oag18 = new_oag18.reset_index()
new_oag18 = new_oag18.drop(columns='index')

**Check whether the reassignment was successful: None of the to-be-reassigned-ac-codes should be in the df anymore:**

In [12]:
ac_code_iata_reassigned = list(reassign_df.ac_code_iata)
ac_code_iata_in_df = list(new_oag18.ac_code_iata.unique())
set(ac_code_iata_in_df).intersection(set(ac_code_iata_reassigned))

set()

**Sanity check:**

Frequency and fkm should be the same for new_oag18 and the entries of oag18:

In [13]:
check_diff_freq = abs(oag18['frequency_2018'].sum() - new_oag18['frequency_2018'].sum())
check_diff_fkm = abs(oag18['fkm_2018'].sum() - new_oag18['fkm_2018'].sum())
print('Difference between frequency before and after ac reassignment: {:.2f}.'.format(check_diff_freq))
print('Difference between total fkm before and after ac reassignment: {:.2f}.'.format(check_diff_fkm))

Difference between frequency before and after ac reassignment: 0.00.
Difference between total fkm before and after ac reassignment: 0.00.


Check passed.

** Since we split the to-be-reassigned ac codes into multiple other ac, we also increased the number of entries: **

In [14]:
len(oag18)

180578

In [15]:
len(new_oag18)

218729

### (2018) Merge duplicate ICAO codes

**Load manual IATA-ICAO lookup table**

In [16]:
# Read the manual lookup table
ac_type_designators = pd.read_excel('../../../../Research/Aviation/Jupyter/datasets/aircraft_type_designators_manual.xlsx')
ac_type_designators = ac_type_designators.rename(columns={'oag_model: all generic AC types where we have multiple variants in the PS data set are marked in red': 'ac_name_oag',
                                            'planespotters_model': 'ac_name_ps',
                                            'share of yearly flight km: high impact models (>1% of total FKM) are marked in green -> ask OAG here': 'share_of_yearly_FKM'})
ac_type_designators = ac_type_designators.drop(['share_of_yearly_FKM','ps_model assignment','comment1','comment2', 'widebody'], axis=1)
ac_type_designators['ac_code_iata_reassign'] = ac_type_designators['ac_code_iata_reassign'].str.split(',')
ac_type_designators['ac_code_iata'] = ac_type_designators['ac_code_iata'].astype(str)
ac_type_designators.head()

ac_code_iata ac_code_icao                 ac_name_oag       ac_name_ps  \
0          CS1         BCS1             Airbus A220-100  Airbus A220-100   
1          CS3         BCS3             Airbus A220-300  Airbus A220-300   
2          AB3         A30B       Airbus A300 Passenger    Airbus A300B4   
3          AB6         A306   Airbus A300-600 Passenger  Airbus A300-600   
4          AB4         A30B  Airbus A300B2/B4 Passenger    Airbus A300B4   

  ac_code_iata_reassign               oag_name_keep  
0                   NaN                         NaN  
1                   NaN                         NaN  
2                   NaN                          no  
3                   NaN                         NaN  
4                   NaN  Airbus A300B2/B4 Passenger

**Remove generic IATA codes**

In [17]:
# We delete all generic IATA codes. They are reassigned to specific aircraft 
# variants in the Jupyter Notebook "OAG_Planespotters_Alignment".
ac_type_designators_wo_generic = ac_type_designators[ac_type_designators['ac_code_icao'] != 'to_be_assigned_to_multiple_variants']
ac_type_designators_wo_generic = ac_type_designators_wo_generic.drop(columns='ac_code_iata_reassign')
ac_type_designators_wo_generic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 173 entries, 0 to 189
Data columns (total 5 columns):
ac_code_iata     173 non-null object
ac_code_icao     173 non-null object
ac_name_oag      173 non-null object
ac_name_ps       127 non-null object
oag_name_keep    66 non-null object
dtypes: object(5)
memory usage: 8.1+ KB


**Add ICAO code to OAG data.**

In [18]:
iata_icao_lookup = ac_type_designators_wo_generic.set_index('ac_code_iata').to_dict()['ac_code_icao']
new_oag18['ac_code_icao'] = new_oag18['ac_code_iata'].replace(iata_icao_lookup)

** Merge duplicate ICAO values**

In [19]:
# Define duplicate icao values
icao_dupl = ac_type_designators_wo_generic[ac_type_designators_wo_generic.duplicated(subset ='ac_code_icao', keep=False)].sort_values(by='ac_code_icao')
icao_dupl[6:10]

ac_code_iata ac_code_icao                ac_name_oag       ac_name_ps  \
13          321         A321                Airbus A321  Airbus A321-200   
14          32B         A321    Airbus A321 (Sharklets)  Airbus A321-200   
28          380         A388      Airbus A380 Passenger  Airbus A380-800   
29          388         A388  Airbus A380-800 Passenger  Airbus A380-800   

                               oag_name_keep  
13  Airbus A321 (with and without Sharklets)  
14                                        no  
28                                        no  
29                 Airbus A380-800 Passenger

Example: The A321 are all given the oag name "Airbus A321 (with and without Sharklets)" and the IATA code '321'.

In [20]:
icao_dupl_names = icao_dupl['ac_code_icao'].unique() # unique icao values in the list
icao_dupl_groups = icao_dupl.groupby('ac_code_icao')

for i in icao_dupl_names: # loop over all duplicate icao values
     # Get all entries from icao_dupl for the selected icao code
    _cache = icao_dupl_groups.get_group(i)
    
    # Rename the ac_name_oag to the one given in icao_dupl
    _rename_oag_name = _cache[_cache['oag_name_keep'] != 'no']['oag_name_keep'].item()
    _keep_iata_codes = _cache[_cache['oag_name_keep'] != 'no']['ac_code_iata'].to_list()
    _rename_iata_index = new_oag18[new_oag18['ac_code_iata'].isin(_keep_iata_codes)].index
    new_oag18.at[_rename_iata_index,'ac_name_oag'] = _rename_oag_name
    
    # Rename the ac_code_iata to the first one in the list 
    _del_iata_codes = _cache[_cache['oag_name_keep'] == 'no']['ac_code_iata'].to_list()
    _del_iata_index = new_oag18[new_oag18['ac_code_iata'].isin(_del_iata_codes)].index
    new_oag18.at[_del_iata_index,'ac_code_iata'] = _keep_iata_codes[0]

Now that we set common names for the same ac_code_icao, we will see duplicates for identical ap-pairs. We merge them.

In [21]:
to_be_merged = new_oag18[new_oag18.duplicated(subset = ['ac_code_icao','dep_ap_code', 'arr_ap_code','dep_ap_name', 'arr_ap_name', 'gc_distance'], keep=False)].sort_index()
grouped = to_be_merged.groupby(['ac_code_icao','dep_ap_code', 'arr_ap_code'])

In [22]:
for ac_group in grouped.groups.keys():
    grouped.get_group(ac_group).index[0]
    idx_first_entry = grouped.get_group(ac_group).index[0]
    idx_other_entries = grouped.get_group(ac_group).index[1:]
    freq_sum = grouped.get_group(ac_group).sum()['frequency_2018']
    fkm_sum = grouped.get_group(ac_group).sum()['fkm_2018']

    # We update the summed up freq and fkm and set it as new value for the 1st entry
    new_oag18.at[idx_first_entry, 'frequency_2018'] = freq_sum
    new_oag18.at[idx_first_entry, 'fkm_2018'] = fkm_sum
    
    # We delete all duplicated values except the first one (which is the one with the updated information)
    new_oag18.drop_duplicates(subset=['ac_code_iata','dep_ap_code', 'arr_ap_code','dep_ap_name', 'arr_ap_name', 'gc_distance'], keep='first',inplace=True) 

## (2018) Final processing

In [23]:
new_oag18

ac_code_iata        ac_name_oag arr_ap_code  \
0               100         Fokker 100         ADE   
1               100         Fokker 100         AMS   
2               100         Fokker 100         AMS   
3               100         Fokker 100         AZD   
4               100         Fokker 100         BND   
5               100         Fokker 100         BND   
6               100         Fokker 100         BRU   
7               100         Fokker 100         BRU   
8               100         Fokker 100         FAO   
9               100         Fokker 100         FSC   
10              100         Fokker 100         HIR   
11              100         Fokker 100         INI   
12              100         Fokker 100         JIB   
13              100         Fokker 100         JIB   
14              100         Fokker 100         KIH   
15              100         Fokker 100         KIH   
16              100         Fokker 100         KRK   
17              100         Fokker 100         KSH   
18              100         Fokker 100         MHD   
19              100         Fokker 100         MUC   
20              100         Fokker 100         MUC   
21              100         Fokker 100         NAN   
22              100         Fokker 100         NCE   
23              100         Fokker 100         NTE   
24              100         Fokker 100         RNS   
25              100         Fokker 100         SRY   
26              100         Fokker 100         THR   
27              100         Fokker 100         TLL   
28              100         Fokker 100         TLL   
29              100         Fokker 100         ZAH   
...             ...                ...         ...   
218686          ER3  Embraer RJ135/145         OMA   
218687          ER4       new_reass_ac         OMA   
218689          ER3  Embraer RJ135/145         ORD   
218690          ER4       new_reass_ac         ORD   
218692          ER3  Embraer RJ135/145         ORD   
218693          ER4       new_reass_ac         ORD   
218695          ER3  Embraer RJ135/145         ORD   
218696          ER4       new_reass_ac         ORD   
218698          ER4       new_reass_ac         ORD   
218699          ER3  Embraer RJ135/145         ORD   
218700          ER4       new_reass_ac         ORD   
218702          ER3  Embraer RJ135/145         ORD   
218703          ER4       new_reass_ac         ORD   
218705          ER3  Embraer RJ135/145         PNS   
218706          ER4       new_reass_ac         PNS   
218708          ER4       new_reass_ac         SCE   
218709          ER4       new_reass_ac         SWO   
218710          ER3  Embraer RJ135/145         THR   
218711          ER4       new_reass_ac         THR   
218713          ER3  Embraer RJ135/145         TRC   
218714          ER4       new_reass_ac         TRC   
218716          ER3  Embraer RJ135/145         TRC   
218717          ER4       new_reass_ac         TRC   
218719          ER4       new_reass_ac         TUL   
218720          ER3  Embraer RJ135/145         XNA   
218721          ER4       new_reass_ac         XNA   
218723          ER3  Embraer RJ135/145         YHZ   
218724          ER4       new_reass_ac         YHZ   
218726          ER3  Embraer RJ135/145         YOW   
218727          ER4       new_reass_ac         YOW   

                                    arr_ap_name arr_ap_cty dep_ap_code  \
0                                          Aden         YE         JIB   
1                                     Amsterdam         NL         BRU   
2                                     Amsterdam         NL         MUC   
3                                          Yazd         IR         BND   
4                                  Bandar Abbas         IR         AZD   
5                                  Bandar Abbas         IR         KIH   
6                              Brussels Airport         BE         AMS   
7                              Brussels Airport         BE        

In [24]:
len(new_oag18)

191618

In [25]:
new_oag18.ac_code_iata.unique()

array(['100', '141', '142', '290', '313', '318', '319', '143', '320',
       '321', '32N', '32Q', '332', '333', '717', '343', '345', '346',
       '351', '359', '388', '733', '722', '73L', '734', '735', '339',
       'AT7', '736', '788', '738', 'AT4', 'E95', 'F50', 'L4T', '739',
       'SF3', '73G', 'F70', 'E70', '744', '74H', '752', '753', '762',
       '763', '764', '772', '773', '77L', '77W', '789', '7M8', '781',
       'A32', 'A58', '7M9', 'AN6', 'A81', 'AB4', 'AB6', 'AN4', 'AR8',
       'AR1', 'BE1', 'A38', 'A40', 'ATP', 'BE2', 'BEC', 'BNI', 'CNA',
       'CNF', 'BE4', 'C27', 'CNT', 'CR2', 'CR1', 'CR7', 'CR9', 'DHT',
       'CRK', 'CS1', 'CS3', 'CV5', 'D28', 'D38', 'DH1', 'DH2', 'DH3',
       'DH4', 'DC9', 'CN2', 'CNJ', 'DHC', 'DHP', 'DH7', 'E75', 'E7W',
       'DHL', 'EM2', 'ER3', 'ER4', 'EMB', 'EP3', 'F28', 'FRJ', 'IL9',
       'J31', 'J32', 'J41', 'M82', 'M83', 'M87', 'M88', 'GJ5', 'GRJ',
       'M90', 'MA6', 'PA1', 'PA2', 'PAG', 'PL2', 'M81', 'S20', 'SU9',
       'SWM', 'TU3',

In [26]:
len(new_oag18.ac_code_iata.unique())

136

Finally, we...
- (1) delete the two columns that have been needed for the calculations.
- (2) reset the index.
- (3) change some data types.
- (4) assign the right ac_name_oag for the new reass ac.

In [27]:
#1,2
new_oag18 = new_oag18.drop(columns=['share_of_fkm_2018', 'route']).reset_index()
new_oag18 = new_oag18.drop(columns=['index'])

In [28]:
#3
new_oag18['frequency_2018'] = new_oag18['frequency_2018'].astype(int)
new_oag18['fkm_2018'] = new_oag18['fkm_2018'].astype(float)

In [29]:
#4-1
# update the ac names that have been double icao values previously
idx_oag_name_double_icao = ac_type_designators[(~ac_type_designators['oag_name_keep'].isnull()) & \
                                               (ac_type_designators['oag_name_keep'] != 'no')].index
ac_type_designators.at[idx_oag_name_double_icao,'ac_name_oag'] = \
                                                ac_type_designators.loc[idx_oag_name_double_icao,'oag_name_keep']

In [30]:
#4-2
ac_iata_to_oag_lookup = ac_type_designators.set_index('ac_code_iata').to_dict()['ac_name_oag']
dict(list(ac_iata_to_oag_lookup.items())[10:13]) # preview
new_oag18.at[:,'ac_name_oag'] = new_oag18['ac_code_iata'].replace(ac_iata_to_oag_lookup)
new_oag18.head()

ac_code_iata ac_name_oag arr_ap_code   arr_ap_name arr_ap_cty dep_ap_code  \
0          100  Fokker 100         ADE          Aden         YE         JIB   
1          100  Fokker 100         AMS     Amsterdam         NL         BRU   
2          100  Fokker 100         AMS     Amsterdam         NL         MUC   
3          100  Fokker 100         AZD          Yazd         IR         BND   
4          100  Fokker 100         BND  Bandar Abbas         IR         AZD   

                    dep_ap_name dep_ap_cty  gc_distance  frequency_2018  \
0                      Djibouti         DJ   248.203212              58   
1              Brussels Airport         BE   157.553624              88   
2  Munich International Airport         DE   676.777570             332   
3                  Bandar Abbas         IR   559.459639              91   
4                          Yazd         IR   559.459639              91   

        fkm_2018 ac_code_icao  
0   14395.786303         F100  
1   13864.718918         F100  
2  224690.153347         F100  
3   50910.827163         F100  
4   50910.827163         F100

In [31]:
#4: Check whether all oag names have been replaced
len(new_oag18[new_oag18['ac_name_oag'] == 'new_reass_ac'])

0

In [32]:
# Delete ICAO code
new_oag18 = new_oag18.drop(columns='ac_code_icao')

In [33]:
new_oag18.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191618 entries, 0 to 191617
Data columns (total 11 columns):
ac_code_iata      191618 non-null object
ac_name_oag       191618 non-null object
arr_ap_code       191618 non-null object
arr_ap_name       191618 non-null object
arr_ap_cty        191618 non-null object
dep_ap_code       191618 non-null object
dep_ap_name       191618 non-null object
dep_ap_cty        191618 non-null object
gc_distance       191618 non-null float64
frequency_2018    191618 non-null int32
fkm_2018          191618 non-null float64
dtypes: float64(2), int32(1), object(8)
memory usage: 15.4+ MB


In [34]:
print('By reassigning of aircraft, the number of entries has increased by {:.2f}% from {} to {} entries.'.format(len(new_oag18)/len(oag18)*100-100,len(oag18), len(new_oag18)))

By reassigning of aircraft, the number of entries has increased by 6.11% from 180578 to 191618 entries.


In [36]:
new_oag18.to_csv('intermediary_save_of_new_oag18.csv')

## (2018) Upload changed OAG data set

In [35]:
# Checking the maximum character length in columns:
new_oag18.ac_name_oag.str.len().max()

64

In [37]:
%pg_sql set role "NDA-OAG-full"

SUCCESS: query did not return any data


<cursor object at 0x000001E2E7B53E48; closed: 0>

In [38]:
%%pg_sql cur
drop table if exists "OAG"."2018_schedule_ps_conform";

CREATE TABLE "OAG"."2018_schedule_ps_conform" 
("ac_code_iata" varchar(3), 
 "ac_name_oag" varchar(64), 
 "arr_ap_code" varchar(3),
 "arr_ap_name" varchar(39),
 "arr_ap_cty" varchar(2),
 "dep_ap_code" varchar(3),
 "dep_ap_name" varchar(39),
 "dep_ap_cty" varchar(2),
 "gc_distance" float,
 "frequency_2018" int,
 "fkm_2018" float
);

COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."ac_code_iata" IS 'IATA aircraft code';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."ac_name_oag" IS 'OAG aircraft name';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."arr_ap_code" IS 'IATA aiport code';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."arr_ap_name" IS 'IATA aiport name';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."arr_ap_cty" IS  'IATA airport country';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."dep_ap_code" IS 'IATA aiport code';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."dep_ap_name" IS 'IATA aiport name';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."dep_ap_cty" IS 'IATA airport country';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."gc_distance" IS 'great-circle distance for this segment';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."frequency_2018" IS 'number of flights for this segment in 2018';
COMMENT ON
COLUMN "OAG"."2018_schedule_ps_conform"."fkm_2018" IS 'flight kilometers for this segment (gc_distance * frequency)';

COMMIT;

SUCCESS: query did not return any data
 cursor object as 'cur'


In [39]:
%pg_copy new_oag18 OAG.2018_schedule_ps_conform

  waring: green-mode temporarily deactivated (interrupt won't abort the import)  green mode reactivated

In [40]:
%pg_disconnect

# 2012

In [47]:
# Connect to DB
%load_ext ipython_pg.LAV
%pg_connect dbname='mobility' host='lav-fileserver.ethz.ch' user='heldm'

The ipython_pg.LAV extension is already loaded. To reload it, use:
  %reload_ext ipython_pg.LAV
password for heldm@lav-fileserver.ethz.ch:5433:········
SUCCESS: connected to lav-fileserver.ethz.ch
  PostGIS integration enabled

In [48]:
%%pg_pd oag12
SELECT * FROM "OAG"."2012_schedule"

SUCCES: matched 135894 rows
 results stored as 'oag12'


In [49]:
oag12.head()

ac_code_iata ac_name_oag arr_ap_code         arr_ap_name arr_ap_cty  \
0          100  Fokker 100         AGT     Ciudad del Este         PY   
1          100  Fokker 100         AMS           Amsterdam         NL   
2          100  Fokker 100         ASU            Asuncion         PY   
3          100  Fokker 100         BHX  Birmingham Airport         GB   
4          100  Fokker 100         BIO              Bilbao         ES   

  dep_ap_code      dep_ap_name dep_ap_cty  gc_distance  frequency_2012  \
0         ASU         Asuncion         PY   303.201008           176.0   
1         LUX       Luxembourg         LU   314.611383             1.0   
2         AGT  Ciudad del Este         PY   303.201008           225.0   
3         ZRH   Zurich Airport         CH   921.176839           668.0   
4         LIS           Lisbon         PT   723.541703             1.0   

        fkm_2012  
0   53363.377470  
1     314.611383  
2   68220.226879  
3  615346.128185  
4     723.541703

## (2012) Make OAG data set Planespotters-compatible

### (2012) Reassign generic IATA Codes

**Load manual IATA-ICAO lookup table**

In [50]:
# Read the manual lookup table
ac_type_designators = pd.read_excel('../../../../Research/Aviation/Jupyter/datasets/aircraft_type_designators_manual.xlsx')
ac_type_designators = ac_type_designators.rename(columns={'oag_model: all generic AC types where we have multiple variants in the PS data set are marked in red': 'ac_name_oag',
                                            'planespotters_model': 'ac_name_ps',
                                            'share of yearly flight km: high impact models (>1% of total FKM) are marked in green -> ask OAG here': 'share_of_yearly_FKM'})
ac_type_designators = ac_type_designators.drop(['share_of_yearly_FKM','ps_model assignment','comment1','comment2', 'widebody'], axis=1)
ac_type_designators['ac_code_iata_reassign'] = ac_type_designators['ac_code_iata_reassign'].str.split(',')
ac_type_designators['ac_code_iata'] = ac_type_designators['ac_code_iata'].astype(str)
ac_type_designators.head()

ac_code_iata ac_code_icao                 ac_name_oag       ac_name_ps  \
0          CS1         BCS1             Airbus A220-100  Airbus A220-100   
1          CS3         BCS3             Airbus A220-300  Airbus A220-300   
2          AB3         A30B       Airbus A300 Passenger    Airbus A300B4   
3          AB6         A306   Airbus A300-600 Passenger  Airbus A300-600   
4          AB4         A30B  Airbus A300B2/B4 Passenger    Airbus A300B4   

  ac_code_iata_reassign               oag_name_keep  
0                   NaN                         NaN  
1                   NaN                         NaN  
2                   NaN                          no  
3                   NaN                         NaN  
4                   NaN  Airbus A300B2/B4 Passenger

In [51]:
# Extract all iata codes that have to be reassigned
reassign_df = ac_type_designators[ac_type_designators['ac_name_ps'] == 'to_be_assigned_to_multiple_variants'].reset_index()
reassign_df

index ac_code_iata                         ac_code_icao  \
0       8          32S  to_be_assigned_to_multiple_variants   
1      16          330  to_be_assigned_to_multiple_variants   
2      20          340  to_be_assigned_to_multiple_variants   
3      25          350  to_be_assigned_to_multiple_variants   
4      40          ATR  to_be_assigned_to_multiple_variants   
5      42          ARJ  to_be_assigned_to_multiple_variants   
6      44          146  to_be_assigned_to_multiple_variants   
7      62          M80  to_be_assigned_to_multiple_variants   
8      73          737  to_be_assigned_to_multiple_variants   
9      98          757  to_be_assigned_to_multiple_variants   
10    103          767  to_be_assigned_to_multiple_variants   
11    108          777  to_be_assigned_to_multiple_variants   
12    113          787  to_be_assigned_to_multiple_variants   
13    120          CRJ  to_be_assigned_to_multiple_variants   
14    141          DH8  to_be_assigned_to_multiple_variants   
15    149          EMJ  to_be_assigned_to_multiple_variants   
16    156          ERJ  to_be_assigned_to_multiple_variants   

                      ac_name_oag                           ac_name_ps  \
0         Airbus A318/319/320/321  to_be_assigned_to_multiple_variants   
1                     Airbus A330  to_be_assigned_to_multiple_variants   
2                     Airbus A340  to_be_assigned_to_multiple_variants   
3                     Airbus A350  to_be_assigned_to_multiple_variants   
4                    ATR42 /ATR72  to_be_assigned_to_multiple_variants   
5          Avro RJ70 /rj85 /rj100  to_be_assigned_to_multiple_variants   
6               BAe 146 Passenger  to_be_assigned_to_multiple_variants   
7          Boeing (douglas) MD-80  to_be_assigned_to_multiple_variants   
8            Boeing 737 Passenger  to_be_assigned_to_multiple_variants   
9          Boeing 757 (Passenger)  to_be_assigned_to_multiple_variants   
10           Boeing 767 Passenger  to_be_assigned_to_multiple_variants   
11           Boeing 777 Passenger  to_be_assigned_to_multiple_variants   
12                     Boeing 787  to_be_assigned_to_multiple_variants   
13          Canadair Regional Jet  to_be_assigned_to_multiple_variants   
14  DHvilld-Bombardier DHC8 Dsh 8  to_be_assigned_to_multiple_variants   
15                Embraer 170/195  to_be_assigned_to_multiple_variants   
16         Embraer RJ 135/140/145  to_be_assigned_to_multiple_variants   

                                ac_code_iata_reassign oag_name_keep  
0                      [318, 319, 320, 32A, 321, 32B]           NaN  
1                                          [332, 333]           NaN  
2                                [342, 343, 345, 346]           NaN  
3                                          [351, 359]           NaN  
4                                          [AT4, AT7]           NaN  
5                                          [AR1, AR8]           NaN  
6                                     [141, 142, 143]           NaN  
7                           [M81, M82, M83, M87, M88]           NaN  
8   [732, 73C, 733, 734, 73E, 735, 736, 73W, 73G, ...           NaN  
9                                          [752, 753]           NaN  
10                                    [762, 763, 764]           NaN  
11                                         [772, 773]           NaN  
12                                    [781, 788, 789]           NaN  
13                     [CR1, CR2, CR7, CR9, CRA, CRK]           NaN  
14                               [DH1, DH2, DH3, DH4]           NaN  
15                                         [E70, E95]           NaN  
16                                    [ER3, ER4, ERD]           NaN

In [52]:
# Some of these aircraft (that we have assigned manually for the aircraft fleet in 2018) have not been in service yet
# in 2012. Hence, we filter out all aircraft that are in the to-be-reassigned-list, but not present in the 2012-fleet.
_ls = list(reassign_df.loc[:,'ac_code_iata_reassign'])
flat_ls_reassign = [item for sublist in _ls for item in sublist]
flat_ls_oag12 = list(oag12.ac_code_iata.unique())
to_be_del = list(set(flat_ls_reassign)-set(flat_ls_oag12)) # Returns all elements of flat_ls_reassign that are not in flat_ls_oag12
to_be_del

['789', '781', '351', '32B', '359', '7S8']

In [53]:
# The '350' has not yet been in service at all (-> hence neither the '359' not the '351') 
# -> we delete that row from the reass df
reassign_df = reassign_df.drop(index=reassign_df[reassign_df.ac_code_iata=='350'].index).reset_index()
reassign_df = reassign_df.drop(columns = 'level_0')

In [54]:
# For the others, we filter the new aircraft types (see to_be_del list)
for i in np.arange(len(reassign_df)):
    cur_ac = reassign_df.ac_code_iata[i]
    intersect = set(reassign_df.ac_code_iata_reassign[i]).intersection(set(to_be_del))
    if len(intersect) > 0:
        print('Old list for {}: {}'.format(cur_ac, reassign_df.iloc[i,5]))
        print('New list for {}: {}'.format(cur_ac, list(set(reassign_df.ac_code_iata_reassign[i]) - (set(to_be_del)))))
        reassign_df.at[i,'ac_code_iata_reassign'] = list(set(reassign_df.ac_code_iata_reassign[i]) - (set(to_be_del)))

Old list for 32S: ['318', '319', '320', '32A', '321', '32B']
New list for 32S: ['319', '32A', '321', '320', '318']
Old list for 737: ['732', '73C', '733', '734', '73E', '735', '736', '73W', '73G', '7S8', '73H', '738', '73J', '739']
New list for 737: ['734', '736', '738', '73J', '735', '73W', '73H', '73G', '733', '739', '73C', '732', '73E']
Old list for 787: ['781', '788', '789']
New list for 787: ['788']


In [55]:
reassign_df

index ac_code_iata                         ac_code_icao  \
0       8          32S  to_be_assigned_to_multiple_variants   
1      16          330  to_be_assigned_to_multiple_variants   
2      20          340  to_be_assigned_to_multiple_variants   
3      40          ATR  to_be_assigned_to_multiple_variants   
4      42          ARJ  to_be_assigned_to_multiple_variants   
5      44          146  to_be_assigned_to_multiple_variants   
6      62          M80  to_be_assigned_to_multiple_variants   
7      73          737  to_be_assigned_to_multiple_variants   
8      98          757  to_be_assigned_to_multiple_variants   
9     103          767  to_be_assigned_to_multiple_variants   
10    108          777  to_be_assigned_to_multiple_variants   
11    113          787  to_be_assigned_to_multiple_variants   
12    120          CRJ  to_be_assigned_to_multiple_variants   
13    141          DH8  to_be_assigned_to_multiple_variants   
14    149          EMJ  to_be_assigned_to_multiple_variants   
15    156          ERJ  to_be_assigned_to_multiple_variants   

                      ac_name_oag                           ac_name_ps  \
0         Airbus A318/319/320/321  to_be_assigned_to_multiple_variants   
1                     Airbus A330  to_be_assigned_to_multiple_variants   
2                     Airbus A340  to_be_assigned_to_multiple_variants   
3                    ATR42 /ATR72  to_be_assigned_to_multiple_variants   
4          Avro RJ70 /rj85 /rj100  to_be_assigned_to_multiple_variants   
5               BAe 146 Passenger  to_be_assigned_to_multiple_variants   
6          Boeing (douglas) MD-80  to_be_assigned_to_multiple_variants   
7            Boeing 737 Passenger  to_be_assigned_to_multiple_variants   
8          Boeing 757 (Passenger)  to_be_assigned_to_multiple_variants   
9            Boeing 767 Passenger  to_be_assigned_to_multiple_variants   
10           Boeing 777 Passenger  to_be_assigned_to_multiple_variants   
11                     Boeing 787  to_be_assigned_to_multiple_variants   
12          Canadair Regional Jet  to_be_assigned_to_multiple_variants   
13  DHvilld-Bombardier DHC8 Dsh 8  to_be_assigned_to_multiple_variants   
14                Embraer 170/195  to_be_assigned_to_multiple_variants   
15         Embraer RJ 135/140/145  to_be_assigned_to_multiple_variants   

                                ac_code_iata_reassign oag_name_keep  
0                           [319, 32A, 321, 320, 318]           NaN  
1                                          [332, 333]           NaN  
2                                [342, 343, 345, 346]           NaN  
3                                          [AT4, AT7]           NaN  
4                                          [AR1, AR8]           NaN  
5                                     [141, 142, 143]           NaN  
6                           [M81, M82, M83, M87, M88]           NaN  
7   [734, 736, 738, 73J, 735, 73W, 73H, 73G, 733, ...           NaN  
8                                          [752, 753]           NaN  
9                                     [762, 763, 764]           NaN  
10                                         [772, 773]           NaN  
11                                              [788]           NaN  
12                     [CR1, CR2, CR7, CR9, CRA, CRK]           NaN  
13                               [DH1, DH2, DH3, DH4]           NaN  
14                                         [E70, E95]           NaN  
15                                    [ER3, ER4, ERD]           NaN

**Reassign**

In [56]:
# Add the total share of yearly fkm of each OAG entry and a one-string-route as two new columns
oag12['share_of_fkm_2012'] = oag12['fkm_2012'] / oag12['fkm_2012'].sum()
oag12['route'] = oag12['arr_ap_code'] + '_' + oag12['dep_ap_code']
oag12.head()

ac_code_iata ac_name_oag arr_ap_code         arr_ap_name arr_ap_cty  \
0          100  Fokker 100         AGT     Ciudad del Este         PY   
1          100  Fokker 100         AMS           Amsterdam         NL   
2          100  Fokker 100         ASU            Asuncion         PY   
3          100  Fokker 100         BHX  Birmingham Airport         GB   
4          100  Fokker 100         BIO              Bilbao         ES   

  dep_ap_code      dep_ap_name dep_ap_cty  gc_distance  frequency_2012  \
0         ASU         Asuncion         PY   303.201008           176.0   
1         LUX       Luxembourg         LU   314.611383             1.0   
2         AGT  Ciudad del Este         PY   303.201008           225.0   
3         ZRH   Zurich Airport         CH   921.176839           668.0   
4         LIS           Lisbon         PT   723.541703             1.0   

        fkm_2012  share_of_fkm_2012    route  
0   53363.377470       1.372065e-06  AGT_ASU  
1     314.611383       8.089205e-09  AMS_LUX  
2   68220.226879       1.754061e-06  ASU_AGT  
3  615346.128185       1.582162e-05  BHX_ZRH  
4     723.541703       1.860351e-08  BIO_LIS

In [57]:
def lookup_fkm_shares(iata_orig, ls_reassign, oag12_route, oag12):
    '''Take the original generic iata code aircraft (iata_orig) that has to be reassigned to other aircraft types,
    which are the input list ls_reassign aircraft types.
    Takes the oag data for a specific route - all aircraft that perform these route.
    Returns relative shares of these to-be-reassigned aircraft types.'''
    
    ls_reassign = pd.DataFrame({'ac_code_iata': ls_reassign, 'share_of_fkm_2012': 'share_of_fkm_2012'})
    ac_performance = oag12.groupby('ac_code_iata').sum().reset_index()
    ls_reassign['share_of_fkm_2012'] = ac_performance.set_index('ac_code_iata').lookup(ls_reassign['ac_code_iata'],ls_reassign['share_of_fkm_2012'])
    ls_reassign['share_of_fkm_2012'] = ls_reassign['share_of_fkm_2012'] / ls_reassign['share_of_fkm_2012'].sum()
    
    return ls_reassign

In [58]:
new_oag12 = oag12.copy()

In [59]:
'''Some iata codes in the OAG data set do not refer to a specific aircraft, but are a general "melting pot" for all aircraft versions of a general type.
Since stock data from Planespotters does refer only to specific aircraft, we have to reassign the general aircraft types to the specific ones.
This code block assures that this reassignment (redistribution of number of flights (frequency_2012) and flight kilometers (fkm_2012)) is done according to 
the share of specific aircraft versions on the total fkm in 2012. Since fkm correlate good with number of aircraft in a first order approximation,
yearly fkm share is a good proxy for the number of aircraft version that we reassign.'''

for iata_orig in np.arange(len(reassign_df)): # loop over all iata codes that have to be replaced (iata_orig = [0,1,2,...,#all orig iata codes])
    
    orig = reassign_df.loc[iata_orig,'ac_code_iata']           # original ac that has to be reassigned (e.g. '330')
    reass = reassign_df.loc[iata_orig,'ac_code_iata_reassign'] # ac to which the fkm are reassigned (e.g. ['332', '333'])
    
    
    for route in oag12[oag12['ac_code_iata'] == orig]['route']: # loop over routes
        
        if len(oag12[(oag12['route'] == route) & (oag12['ac_code_iata'] == orig)]) > 0: # check if orig ac is operated at the current route
            # Look sequentially at routes in the oag data for the orig aircraft
            oag12_route = oag12[(oag12['route'] == route) & \
                                ((oag12['ac_code_iata'] == orig) | (oag12['ac_code_iata'].isin(reass)))].reset_index()
            oag12_route = oag12_route.drop(columns='index')

            if len(oag12_route) > 0:
                # Create a new table for all reass aircraft and set empty cells that are filled afterwards
                orig_oag12_route = oag12_route[oag12_route['ac_code_iata'] == orig]
                new_df = pd.concat([orig_oag12_route]*len(reass), ignore_index=True)
                new_df.at[:,['ac_code_iata', 'ac_name_oag','frequency_2012', 'fkm_2012', 'share_of_fkm_2012']] = ''
                new_df.at[:,'ac_code_iata'] = reass
                new_df.at[:,'ac_name_oag'] = 'new_reass_ac'

                # Define shares of yearly fkm shares of the reass aircraft
                shares = lookup_fkm_shares(orig, reass, oag12_route, oag12)

                # Calculate the fkm that have to be redistributed from the orig ac to the reass ac
                orig_fkm = oag12_route[oag12_route['ac_code_iata'] == orig]['fkm_2012'].item()


                # Loop over all reass ac for the specific aircraft orig (loop 1) and the specific route (loop 2)
                for i in reass:
                    # Get the share of fkm of i-th reass ac on the total fkm of all reass ac
                    share_i = shares[shares['ac_code_iata'] == i]['share_of_fkm_2012'].item()
            
                    # Check whether the i-th reass aircraft is already assigned for that route in the original oag data set and save the fkm of it if it is nonzero.
                    reass_fkm_old = oag12_route[oag12_route['ac_code_iata'] == i]['fkm_2012']
                    if len(reass_fkm_old) >0:
                        reass_fkm_old = reass_fkm_old.item()
                    else:
                        reass_fkm_old = 0
                    
                    # Fill the dataframe for the reass aircraft with the share-wise distribution of the to-be-reassigned orig fkm + the reass_fkm_old
                    row = new_df[new_df['ac_code_iata'] == i].index.item()
                    
                    
                    new_df.at[row,'fkm_2012'] = share_i*orig_fkm + reass_fkm_old
                    # Calculate the frequency from the fkm and the great-circle distance
                    new_df.at[row,'frequency_2012'] = round(new_df.loc[row,'fkm_2012'] / new_df.loc[row,'gc_distance'])   


                # Since we assign the fkm according to yearyl fkm shares, the frequency for a specific route-ac pair might not be a multiple of the gc_distance (see last step in reass-loop).
                # Solution: Make sure that the sum of frequency_2012 over all reass ac is the same as for all ac (orig + reass) on the route.
                diff_freq = new_df['frequency_2012'].sum() - oag12_route['frequency_2012'].sum() 

                # If there is a difference, add or substract that difference in number of flights from the reass ac with the largest freq,
                # unless there is no reass ac with a largest freq, then take the last ac in the list
                idx_max = pd.to_numeric(new_df['frequency_2012'])

                if abs(diff_freq) > 0:
                    if math.isnan(idx_max.idxmax()):
                        last_row = new_df.index[-1]
                        new_df.at[last_row,'frequency_2012'] -= diff_freq
                    else:
                        new_df.at[idx_max.idxmax(),'frequency_2012'] -= diff_freq

                # After this correction, we can reset the fkm's according to the frequencies:
                new_df.at[:,'fkm_2012'] = new_df.loc[:,'frequency_2012'] * new_df.loc[:,'gc_distance']

                # Now we append the rows with the new values for the reass ac and drop the original ones if existing
                new_oag12 = new_oag12.append(new_df, ignore_index=True)
                new_oag12 = new_oag12.drop_duplicates(subset = ['ac_code_iata','route'], keep = 'last')
                
    # Last we delete all orig ac
    idx_to_del = new_oag12[new_oag12['ac_code_iata'] == orig].index
    new_oag12 = new_oag12.drop(index=idx_to_del)

In [60]:
# Since the fkm-frequency-step in the loop above creates a few 0-frequency-entries, we delete them.
zeros = new_oag12[new_oag12['frequency_2012'] == 0].index
new_oag12 = new_oag12.drop(index=zeros)
new_oag12 = new_oag12.reset_index()
new_oag12 = new_oag12.drop(columns='index')

**Check whether the reassignment was successful: None of the to-be-reassigned-ac-codes should be in the df anymore:**

In [61]:
ac_code_iata_reassigned = list(reassign_df.ac_code_iata)
ac_code_iata_in_df = list(new_oag12.ac_code_iata.unique())
set(ac_code_iata_in_df).intersection(set(ac_code_iata_reassigned))

set()

**Sanity check:**

Frequency and fkm should be the same for new_oag12 and the entries of oag12:

In [62]:
check_diff_freq = abs(oag12['frequency_2012'].sum() - new_oag12['frequency_2012'].sum())
check_diff_fkm = abs(oag12['fkm_2012'].sum() - new_oag12['fkm_2012'].sum())
print('Difference between frequency before and after ac reassignment: {:.2f}.'.format(check_diff_freq))
print('Difference between total fkm before and after ac reassignment: {:.2f}.'.format(check_diff_fkm))

Difference between frequency before and after ac reassignment: 0.00.
Difference between total fkm before and after ac reassignment: 0.00.


Check passed.

** Since we split the to-be-reassigned ac codes into multiple other ac, we also increased the number of entries: **

In [63]:
len(oag12)

135894

In [64]:
len(new_oag12)

186040

### (2012) Merge duplicate ICAO codes

**Load manual IATA-ICAO lookup table**

In [65]:
# Read the manual lookup table
ac_type_designators = pd.read_excel('../../../../Research/Aviation/Jupyter/datasets/aircraft_type_designators_manual.xlsx')
ac_type_designators = ac_type_designators.rename(columns={'oag_model: all generic AC types where we have multiple variants in the PS data set are marked in red': 'ac_name_oag',
                                            'planespotters_model': 'ac_name_ps',
                                            'share of yearly flight km: high impact models (>1% of total FKM) are marked in green -> ask OAG here': 'share_of_yearly_FKM'})
ac_type_designators = ac_type_designators.drop(['share_of_yearly_FKM','ps_model assignment','comment1','comment2', 'widebody'], axis=1)
ac_type_designators['ac_code_iata_reassign'] = ac_type_designators['ac_code_iata_reassign'].str.split(',')
ac_type_designators['ac_code_iata'] = ac_type_designators['ac_code_iata'].astype(str)
ac_type_designators.head()

ac_code_iata ac_code_icao                 ac_name_oag       ac_name_ps  \
0          CS1         BCS1             Airbus A220-100  Airbus A220-100   
1          CS3         BCS3             Airbus A220-300  Airbus A220-300   
2          AB3         A30B       Airbus A300 Passenger    Airbus A300B4   
3          AB6         A306   Airbus A300-600 Passenger  Airbus A300-600   
4          AB4         A30B  Airbus A300B2/B4 Passenger    Airbus A300B4   

  ac_code_iata_reassign               oag_name_keep  
0                   NaN                         NaN  
1                   NaN                         NaN  
2                   NaN                          no  
3                   NaN                         NaN  
4                   NaN  Airbus A300B2/B4 Passenger

**Remove generic IATA codes**

In [66]:
# We delete all generic IATA codes. They are reassigned to specific aircraft 
# variants in the Jupyter Notebook "OAG_Planespotters_Alignment".
ac_type_designators_wo_generic = ac_type_designators[ac_type_designators['ac_code_icao'] != 'to_be_assigned_to_multiple_variants']
ac_type_designators_wo_generic = ac_type_designators_wo_generic.drop(columns='ac_code_iata_reassign')
ac_type_designators_wo_generic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 173 entries, 0 to 189
Data columns (total 5 columns):
ac_code_iata     173 non-null object
ac_code_icao     173 non-null object
ac_name_oag      173 non-null object
ac_name_ps       127 non-null object
oag_name_keep    66 non-null object
dtypes: object(5)
memory usage: 8.1+ KB


**Add ICAO code to OAG data.**

In [67]:
iata_icao_lookup = ac_type_designators_wo_generic.set_index('ac_code_iata').to_dict()['ac_code_icao']
new_oag12['ac_code_icao'] = new_oag12['ac_code_iata'].replace(iata_icao_lookup)

** Merge duplicate ICAO values**

In [68]:
# Define duplicate icao values
icao_dupl = ac_type_designators_wo_generic[ac_type_designators_wo_generic.duplicated(subset ='ac_code_icao', keep=False)].sort_values(by='ac_code_icao')
icao_dupl[6:10]

ac_code_iata ac_code_icao                ac_name_oag       ac_name_ps  \
13          321         A321                Airbus A321  Airbus A321-200   
14          32B         A321    Airbus A321 (Sharklets)  Airbus A321-200   
28          380         A388      Airbus A380 Passenger  Airbus A380-800   
29          388         A388  Airbus A380-800 Passenger  Airbus A380-800   

                               oag_name_keep  
13  Airbus A321 (with and without Sharklets)  
14                                        no  
28                                        no  
29                 Airbus A380-800 Passenger

Example: The A321 are all given the oag name "Airbus A321 (with and without Sharklets)" and the IATA code '321'.

In [69]:
icao_dupl_names = icao_dupl['ac_code_icao'].unique() # unique icao values in the list
icao_dupl_groups = icao_dupl.groupby('ac_code_icao')

for i in icao_dupl_names: # loop over all duplicate icao values
     # Get all entries from icao_dupl for the selected icao code
    _cache = icao_dupl_groups.get_group(i)
    
    # Rename the ac_name_oag to the one given in icao_dupl
    _rename_oag_name = _cache[_cache['oag_name_keep'] != 'no']['oag_name_keep'].item()
    _keep_iata_codes = _cache[_cache['oag_name_keep'] != 'no']['ac_code_iata'].to_list()
    _rename_iata_index = new_oag12[new_oag12['ac_code_iata'].isin(_keep_iata_codes)].index
    new_oag12.at[_rename_iata_index,'ac_name_oag'] = _rename_oag_name
    
    # Rename the ac_code_iata to the first one in the list 
    _del_iata_codes = _cache[_cache['oag_name_keep'] == 'no']['ac_code_iata'].to_list()
    _del_iata_index = new_oag12[new_oag12['ac_code_iata'].isin(_del_iata_codes)].index
    new_oag12.at[_del_iata_index,'ac_code_iata'] = _keep_iata_codes[0]

Now that we set common names for the same ac_code_icao, we will see duplicates for identical ap-pairs. We merge them.

In [70]:
to_be_merged = new_oag12[new_oag12.duplicated(subset = ['ac_code_icao','dep_ap_code', 'arr_ap_code','dep_ap_name', 'arr_ap_name', 'gc_distance'], keep=False)].sort_index()
grouped = to_be_merged.groupby(['ac_code_icao','dep_ap_code', 'arr_ap_code'])

In [71]:
for ac_group in grouped.groups.keys():
    grouped.get_group(ac_group).index[0]
    idx_first_entry = grouped.get_group(ac_group).index[0]
    idx_other_entries = grouped.get_group(ac_group).index[1:]
    freq_sum = grouped.get_group(ac_group).sum()['frequency_2012']
    fkm_sum = grouped.get_group(ac_group).sum()['fkm_2012']

    # We update the summed up freq and fkm and set it as new value for the 1st entry
    new_oag12.at[idx_first_entry, 'frequency_2012'] = freq_sum
    new_oag12.at[idx_first_entry, 'fkm_2012'] = fkm_sum
    
    # We delete all duplicated values except the first one (which is the one with the updated information)
    new_oag12.drop_duplicates(subset=['ac_code_iata','dep_ap_code', 'arr_ap_code','dep_ap_name', 'arr_ap_name', 'gc_distance'], keep='first',inplace=True) 

## (2012) Final processing

In [72]:
new_oag12

ac_code_iata        ac_name_oag arr_ap_code  \
0               100         Fokker 100         AGT   
1               100         Fokker 100         AMS   
2               100         Fokker 100         ASU   
3               100         Fokker 100         BHX   
4               100         Fokker 100         BIO   
5               100         Fokker 100         BNE   
6               100         Fokker 100         BRU   
7               100         Fokker 100         BXR   
8               100         Fokker 100         DPS   
9               100         Fokker 100         FCO   
10              100         Fokker 100         FNC   
11              100         Fokker 100         GRU   
12              100         Fokker 100         GRU   
13              100         Fokker 100         GRU   
14              100         Fokker 100         GYL   
15              100         Fokker 100         LIS   
16              100         Fokker 100         LIS   
17              100         Fokker 100         LIS   
18              100         Fokker 100         LIS   
19              100         Fokker 100         LUX   
20              100         Fokker 100         MAD   
21              100         Fokker 100         MAN   
22              100         Fokker 100         OPO   
23              100         Fokker 100         OPO   
24              100         Fokker 100         OPO   
25              100         Fokker 100         ORY   
26              100         Fokker 100         OTP   
27              100         Fokker 100         PFB   
28              100         Fokker 100         PGU   
29              100         Fokker 100         PNZ   
...             ...                ...         ...   
185996          ER4       new_reass_ac         ORD   
185998          ER3  Embraer RJ135/145         PIB   
185999          ER4       new_reass_ac         PIB   
186001          ER3  Embraer RJ135/145         ROC   
186002          ER4       new_reass_ac         ROC   
186004          ER3  Embraer RJ135/145         RSW   
186005          ER4       new_reass_ac         RSW   
186007          ER3  Embraer RJ135/145         SAV   
186008          ER4       new_reass_ac         SAV   
186010          ER3  Embraer RJ135/145         SHE   
186011          ER4       new_reass_ac         SHE   
186013          ER3  Embraer RJ135/145         SHV   
186014          ER4       new_reass_ac         SHV   
186016          ER3  Embraer RJ135/145         STC   
186017          ER4       new_reass_ac         STC   
186019          ER4       new_reass_ac         SWO   
186020          ER3  Embraer RJ135/145         TGO   
186021          ER4       new_reass_ac         TGO   
186023          ER4       new_reass_ac         TLS   
186024          ER3       new_reass_ac         TLS   
186025          ER3  Embraer RJ135/145         VER   
186026          ER4       new_reass_ac         VER   
186028          ER3  Embraer RJ135/145         XIY   
186029          ER4       new_reass_ac         XIY   
186031          ER3  Embraer RJ135/145         XNA   
186032          ER4       new_reass_ac         XNA   
186034          ER3  Embraer RJ135/145         XNA   
186035          ER4       new_reass_ac         XNA   
186037          ER3  Embraer RJ135/145         YOW   
186038          ER4       new_reass_ac         YOW   

                                    arr_ap_name arr_ap_cty dep_ap_code  \
0                               Ciudad del Este         PY         ASU   
1                                     Amsterdam         NL         LUX   
2                                      Asuncion         PY         AGT   
3                            Birmingham Airport         GB         ZRH   
4                                        Bilbao         ES         LIS   
5                                      Brisbane         AU         POM   
6                              Brussels Airport         BE         OPO   
7                                           Bam         IR        

In [73]:
len(new_oag12)

168547

In [74]:
new_oag12.ac_code_iata.unique()

array(['100', '141', '142', '143', '313', 'AT7', '318', '319', '320',
       '321', '332', '333', '342', '343', '345', '346', '388', '717',
       '722', '72S', '73L', '733', '734', '735', '736', 'AT4', '738',
       '77L', 'AT5', '290', 'E95', '739', '73G', '73A', '73S', 'AR1',
       '743', '744', '74H', '74L', '74M', '752', '753', '74D', '762',
       '763', '742', 'E70', '764', '772', '773', 'E75', '77W', 'F70',
       '788', 'A40', 'A81', 'AB4', 'AB6', 'AN4', 'AN6', 'AR8', '75M',
       'BE1', 'BEC', 'BNI', 'CNA', 'PL2', 'CNF', 'CR2', 'CR7', 'ATP',
       'CR9', 'CNJ', 'CR1', 'BE2', 'CNT', 'CRK', 'CV5', 'D1C', 'D38',
       'D93', 'D95', 'DC9', 'DH1', 'DH2', 'DH3', 'DH4', 'DH7', 'D28',
       'DHL', 'DHP', 'DHT', 'EM2', 'EMB', 'ER3', 'ER4', 'DHC', 'J41',
       'F24', 'F28', 'F50', 'IL9', 'J31', 'J32', 'JST', 'L4T', 'M11',
       'M81', 'M82', 'M83', 'M87', 'M88', 'M90', 'MA6', 'S20', 'PA1',
       'PA2', 'PAG', 'SF3', 'SH6', 'SU9', 'SWM', 'T20', 'TU3', 'TU5',
       'YK2', 'YN2',

In [75]:
len(new_oag12.ac_code_iata.unique())

140

Finally, we...
- (1) delete the two columns that have been needed for the calculations.
- (2) reset the index.
- (3) change some data types.
- (4) assign the right ac_name_oag for the new reass ac.

In [76]:
#1,2
new_oag12 = new_oag12.drop(columns=['share_of_fkm_2012', 'route']).reset_index()
new_oag12 = new_oag12.drop(columns=['index'])

In [77]:
#3
new_oag12['frequency_2012'] = new_oag12['frequency_2012'].astype(int)
new_oag12['fkm_2012'] = new_oag12['fkm_2012'].astype(float)

In [78]:
#4-1
# update the ac names that have been double icao values previously
idx_oag_name_double_icao = ac_type_designators[(~ac_type_designators['oag_name_keep'].isnull()) & \
                                               (ac_type_designators['oag_name_keep'] != 'no')].index
ac_type_designators.at[idx_oag_name_double_icao,'ac_name_oag'] = \
                                                ac_type_designators.loc[idx_oag_name_double_icao,'oag_name_keep']

In [79]:
#4-2
ac_iata_to_oag_lookup = ac_type_designators.set_index('ac_code_iata').to_dict()['ac_name_oag']
dict(list(ac_iata_to_oag_lookup.items())[10:13]) # preview
new_oag12.at[:,'ac_name_oag'] = new_oag12['ac_code_iata'].replace(ac_iata_to_oag_lookup)
new_oag12.head()

ac_code_iata ac_name_oag arr_ap_code         arr_ap_name arr_ap_cty  \
0          100  Fokker 100         AGT     Ciudad del Este         PY   
1          100  Fokker 100         AMS           Amsterdam         NL   
2          100  Fokker 100         ASU            Asuncion         PY   
3          100  Fokker 100         BHX  Birmingham Airport         GB   
4          100  Fokker 100         BIO              Bilbao         ES   

  dep_ap_code      dep_ap_name dep_ap_cty  gc_distance  frequency_2012  \
0         ASU         Asuncion         PY   303.201008             176   
1         LUX       Luxembourg         LU   314.611383               1   
2         AGT  Ciudad del Este         PY   303.201008             225   
3         ZRH   Zurich Airport         CH   921.176839             668   
4         LIS           Lisbon         PT   723.541703               1   

        fkm_2012 ac_code_icao  
0   53363.377470         F100  
1     314.611383         F100  
2   68220.226879         F100  
3  615346.128185         F100  
4     723.541703         F100

In [80]:
#4: Check whether all oag names have been replaced
len(new_oag12[new_oag12['ac_name_oag'] == 'new_reass_ac'])

0

In [81]:
# Delete ICAO code
new_oag12 = new_oag12.drop(columns='ac_code_icao')

In [82]:
new_oag12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 168547 entries, 0 to 168546
Data columns (total 11 columns):
ac_code_iata      168547 non-null object
ac_name_oag       168547 non-null object
arr_ap_code       168547 non-null object
arr_ap_name       168547 non-null object
arr_ap_cty        168547 non-null object
dep_ap_code       168547 non-null object
dep_ap_name       168547 non-null object
dep_ap_cty        168547 non-null object
gc_distance       168547 non-null float64
frequency_2012    168547 non-null int32
fkm_2012          168547 non-null float64
dtypes: float64(2), int32(1), object(8)
memory usage: 13.5+ MB


In [83]:
print('By reassigning of aircraft, the number of entries has increased by {:.2f}% from {} to {} entries.'.format(len(new_oag12)/len(oag12)*100-100,len(oag12), len(new_oag12)))

By reassigning of aircraft, the number of entries has increased by 24.03% from 135894 to 168547 entries.


## (2012) Upload changed OAG data set

In [84]:
# Checking the maximum character length in columns:
new_oag12.ac_name_oag.str.len().max()

64

In [46]:
%pg_sql set role "NDA-OAG-full"

SUCCESS: query did not return any data


<cursor object at 0x000001C6940A4AC8; closed: 0>

In [47]:
%%pg_sql cur
drop table if exists "OAG"."2012_schedule_ps_conform";

CREATE TABLE "OAG"."2012_schedule_ps_conform" 
("ac_code_iata" varchar(3), 
 "ac_name_oag" varchar(64), 
 "arr_ap_code" varchar(3),
 "arr_ap_name" varchar(39),
 "arr_ap_cty" varchar(2),
 "dep_ap_code" varchar(3),
 "dep_ap_name" varchar(39),
 "dep_ap_cty" varchar(2),
 "gc_distance" float,
 "frequency_2012" int,
 "fkm_2012" float
);

COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."ac_code_iata" IS 'IATA aircraft code';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."ac_name_oag" IS 'OAG aircraft name';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."arr_ap_code" IS 'IATA aiport code';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."arr_ap_name" IS 'IATA aiport name';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."arr_ap_cty" IS  'IATA airport country';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."dep_ap_code" IS 'IATA aiport code';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."dep_ap_name" IS 'IATA aiport name';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."dep_ap_cty" IS 'IATA airport country';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."gc_distance" IS 'great-circle distance for this segment';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."frequency_2012" IS 'number of flights for this segment in 2012';
COMMENT ON
COLUMN "OAG"."2012_schedule_ps_conform"."fkm_2012" IS 'flight kilometers for this segment (gc_distance * frequency)';

COMMIT;

SUCCESS: query did not return any data
 cursor object as 'cur'


In [48]:
%pg_copy new_oag12 OAG.2012_schedule_ps_conform

  waring: green-mode temporarily deactivated (interrupt won't abort the import)  green mode reactivated

In [ ]:
%pg_disconnect